Automattic/pocket-casts-ios
# Reimagine Sharing: Export video

| 📘 Part of: #1910 |
|:---:|

Adds Video export to clip sharing. Note that the asset and loading progress are not final. This is mostly adding the video asset writing code.

https://github.com/user-attachments/assets/57e8a42e-5b54-4bae-a09e-8d38072d190f

#### Video Export process
* Exports a 5 second clip of animated kidney shapes - this roughly follows the design specs but could still use tweaking in the future
* Loops this 5 second clip repeatedly over the length of the clip - this was done to speed up rendering as rendering each frame is quite slow
* Adds the audio from the episode to the clip

#### Progress
Adds a progress bar to verify export progress which reports each frame rendered. This isn't the final design but the current version is mainly meant for debugging to verify parts of the process.


### Code Changes

There are a lot of additions here but you can check separately or mostly ignore the helpers added in:

* `UnsafeTransfer.swift` - Removes compiler warnings about unsafe access to `CVPixelBuffer`
* `UnsafeWait.swift` - Allows synchronous blocking of the async operation in `unsafeRequestMediaDataWhenReady`
* `View+CVPixelBuffer.swift` - Generates `CVPixelBuffer` rendering to its context instead of converting to an image first and create a pixel buffer from that, reducing memory usage during export

The major additions are in
* `VideoExporter` - to export the video file
* `AnimatedShareImageView` - to wrap the `ShareImageView` with progress updates

## To test

> [!NOTE]
> Enable the `newSharing` feature flag to test this

* Play an episode
* Tap the "Share" button in the action shelf
* Tap the "Clip" button
* Set the clip length by adjusting the trim handles
* Tap the "Clip" button
* Verify that export progress appears
* Verify that export completes
* Save audio to Photos or Files or share to an app
* Verify that the resulting file includes the clipped audio & subtle animation

## Checklist
